File: MachineLearning/hw02/affprop.cpp
#include <bits/stdc++.h>
using namespace std;

#define SQR(x) ((x)*(x))

double INF = 1e15;

const int MAXN = 500;
double s[MAXN][MAXN];
double r[MAXN][MAXN];
double a[MAXN][MAXN];
double tmp_r[MAXN][MAXN];
double tmp_a[MAXN][MAXN];

struct affinity_propagation
{

    affinity_propagation(vector<double> const& x_coord, 
            vector<double> const& y_coord, int max_iter_,
            int consecutive_max_unchanged_)
        : n(x_coord.size())
        , max_iter(max_iter_)
        , consecutive_max_unchanged(consecutive_max_unchanged_)
    {

        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++) 
            {
                s[i][j] = metric_squared_error(make_pair(x_coord[i], y_coord[i]),
                        make_pair(x_coord[j], y_coord[j]));
                cout << i << " " << j << " " << s[i][j] << endl;
            }

        for (int i = 0; i < n; i++)
        {
            s[i][i] = -1.0;
        }

        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
            {
                r[i][j] = a[i][j] = 0;
            }
    }


    vector<int> get_clusters()
    {
        auto clusters = calc_clusters();

        int it = 0;
        int unchanged = 0;
        while (unchanged < consecutive_max_unchanged and it < max_iter)
        {
            ++it;
            cout << "ITERATION " << it << endl;

            update_values();

            auto new_clusters = calc_clusters();

            if (new_clusters == clusters)
            {
                unchanged++;
            } 
            else 
            {
                unchanged = 1;
            }
        }

        return clusters;
    }

private:

    void print_r()
    {
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n; j++)
            {
                cout << r[i][j] << " \n"[j==n-1];
            }
        }
        cout << endl;
    }

    void update_r()
    {
        for (int i = 0; i < n; i++)
        {
            for (int k = 0; k < n; k++)
            {
                double mx_val = -INF;
                for (int m = 0; m < n; m++)
                {
                    if (m != k)
                    {
                        mx_val = max(mx_val, a[i][m] + s[i][m]);
                    }
                }

                r[i][k] = s[i][k] - mx_val;
            }
        }
    }


    void update_a()
    {
        for (int i = 0; i < n; i++)
        {
            for (int k = 0; k < n; k++)
            {
                if (k == i)
                {
                    double sum_vals = 0;
                    for (int j = 0; j < n; j++)
                    {
                        if (j != i) 
                        {
                            sum_vals += max<double>(0, r[j][k]);
                        }
                    }
                    a[i][k] = sum_vals;
                }
                else
                {
                    double sum_vals = 0;
                    for (int j = 0; j < n; j++) 
                    {
                        if (j != k and j != i)
                        {
                            sum_vals += max<double>(0, r[j][k]);
                        }
                    }
                    a[i][k] = min<double>(0, r[k][k] + sum_vals);
                }
            }
        }
    }

    vector<int> calc_clusters()
    {
        map<int, vector<int>> point_to_cluster;

        for (int i = 0; i < n; i++)
        {
            int opt_k = i;
            for (int k = 0; k < n; k++)
            {
                if (false and i == 100) 
                {
                    cout << a[i][k] + r[i][k] << endl;
                }
                if (a[i][k] + r[i][k] > a[i][opt_k] + r[i][opt_k])
                {
                    opt_k = k;
                }
            }

            point_to_cluster[opt_k].push_back(i);
        }

        vector<int> clusters(n, 0);

        int cl_num = 0;
        for (auto const& kv: point_to_cluster)
        {
            if (kv.second.empty())
            {
                continue;
            }
            
            cl_num++;
            for (const auto &u: kv.second)
            {
                clusters[u] = cl_num;
            }
        }

        return  clusters; 
    }

    void update_values()
    {
        update_r();
        update_a();

        /*
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
            {
                if (abs(a[i][j] -tmp_a[i][j]) > 1e-5)
                    cout << i << " "<< j << endl;
                a[i][j] = tmp_a[i][j];
                if (abs(r[i][j] -tmp_r[i][j]) > 1e-5)
                    cout << i << " "<< j << endl;
                r[i][j] = tmp_r[i][j];
            }
            */
    }

    double metric_squared_error(pair<double, double> p1, pair<double, double> p2)
    {
        return -1 * (SQR(p1.first - p2.first) + SQR(p1.second - p2.second));
    }

    const int n;
    const int max_iter;
    const int consecutive_max_unchanged;
};


int main() 
{
    freopen("blobs.csv", "r", stdin);
    string l;
    cin >> l;
    
    vector<double> x_coord, y_coord;

    char c;
    double x, y;
    while (cin >> x >> c >> y) 
    {
        x_coord.push_back(x);
        y_coord.push_back(y);
    }

    auto clusterer = affinity_propagation(x_coord, y_coord, 5, 15);
    auto clusters = clusterer.get_clusters();

    for (auto const& cl: clusters)
        cout << cl << " ";
    return 0;
}

File: MachineLearning/hw02/affprop.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import colors as plt_colors

from sklearn.cluster import AffinityPropagation as AffinityPropagation_SK
from sklearn.metrics.pairwise import euclidean_distances
from scipy.spatial.distance import euclidean

import sys
import os.path
import random
from subprocess import getoutput, getstatusoutput
from collections import defaultdict


RESULT_DIR = './exp_result'


class AffinityPropagation:

    def __init__(self, max_iter=1, max_cons_iter=5, init_self_val=None, damping=0.9):
        self.max_iter = max_iter
        self.max_cons_iter = max_cons_iter
        self.init_self_val = init_self_val
        # See http://www.psi.toronto.edu/affinitypropagation/faq.html for details.
        self.damping = damping

    
    def _calc_clusters(self):
        point_to_cluster = defaultdict(list)

        for i in range(self.n):
            vals = [self.a[i][k] + self.r[i][k] for k in range(self.n)]
            mx_idx = np.argmax(vals)
#            print(i, vals[mx_idx], vals)

            point_to_cluster[mx_idx].append(i)

        clusters = [-1 for i in range(self.n)]
        for idx, best_guy in enumerate(point_to_cluster.keys()):
            for clustered_with_guy in point_to_cluster[best_guy]:
                clusters[clustered_with_guy] = idx

        return clusters, list(point_to_cluster.keys())



    def _update_r(self):
        for i in range(self.n):
            vals = [self.a[i][m] + self.affinity[i][m] for m in range(self.n)]
            ind = np.argpartition(vals, -2)[-2:]

            # ind[0] should be max index
            if vals[ind[0]] < vals[ind[1]]:
                ind[0], ind[1] = ind[1], ind[0]

            assert(ind[0] == np.argmax(vals))
            assert(ind[0] != ind[1])

            for k in range(self.n):
                sub = vals[ind[1]] if k == ind[0] else vals[ind[0]]
                old_val = self.r[i][k]
                new_val = self.affinity[i][k] - sub

                self.r[i][k] = self.damping * old_val + (1 - self.damping) * new_val

        #print('r updated')


    def _update_a(self):
        for k in range(self.n):
            vals = [max(0, self.r[j][k]) for j in range(self.n)]
            full_sum = sum(vals)

            for i in range(self.n):
                old_val = self.a[i][k]
                new_val = None
                if i == k:
                    new_val = full_sum - vals[i]
                else:
                    new_val = min(0, self.r[k][k] + full_sum - vals[i] - vals[k])

                self.a[i][k] = self.damping * old_val + (1 - self.damping) * new_val

        #print(np.matrix(self.a))
        #print('a updated')


    def cluster(self, objects):
        self.n = objects.shape[0]
        self.affinity = [[0 for i in range(self.n)] for j in range(self.n)]

        self.r = [[0 for i in range(self.n)] for j in range(self.n)]
        self.a = [[0 for i in range(self.n)] for j in range(self.n)]


        self.affinity = -euclidean_distances(objects, squared=True)
        if self.init_self_val is None:
            self.init_self_val = np.median(self.affinity)

        for i in range(self.n):
            self.affinity[i][i] = self.init_self_val
        print('Main diagonal affinity is {}'.format(self.init_self_val))

        clusters, centers_idx = self._calc_clusters()

        it = 0
        unchanged = 0
        while it < self.max_iter and unchanged < self.max_cons_iter:
            it += 1
            print('ITERATION {}'.format(it))

            self._update_r()
            self._update_a()

            new_clusters, centers_idx = self._calc_clusters()
            #print(new_clusters)

            if new_clusters == clusters:
                unchanged += 1
            else:
                unchanged = 1

            print('Clusters did not change for {} iterations'.format(unchanged))
            clusters = new_clusters

        return clusters, centers_idx


def read_data():
    return pd.read_csv('blobs.csv')


def main():
    random.seed(42)

    objects = read_data()


    if True:
        clusterer = AffinityPropagation(max_iter=100, max_cons_iter=15, 
                damping=0.7, init_self_val=-300)
        obj_to_cluster, centers_idx = clusterer.cluster(objects)
    else:
        clusterer = AffinityPropagation_SK()
        obj_to_cluster = clusterer.fit_predict(objects)
        centers_idx = clusterer.cluster_centers_indices_

    x_coord = objects['X']
    y_coord = objects['Y']
    color_names = ['red', 'green', 'blue', 'yellow'] + list(sorted(plt_colors.cnames.keys()))[4::8]
    if len(color_names) < len(set(obj_to_cluster)):
        colors = np.array(obj_to_cluster)
    else:
        colors = [color_names[x] for x in obj_to_cluster]

    print('Number of clusters: {}'.format(len(set(obj_to_cluster))))
    plt.figure()
    plt.scatter(x_coord, y_coord, c=colors)
    for center_id in centers_idx:
        plt.scatter([objects.iloc[center_id][0]], [objects.iloc[center_id][1]], marker='<', 
                c=colors[center_id], 
                s=100)
    plt.savefig(os.path.join(RESULT_DIR, 'affprop.png'))


if __name__ == '__main__':
    main()
